mm2src/coins/utxo/rpc_clients/connection_manager_multiple.rs

```diff
@@ -0,0 +1,399 @@
+        // Find script_hash connection/subscription
+        if let Some(connection) = guard.scripthash_subs.clone().get(script_hash) {
+            let connection = connection.lock().await;
+            // return true if there's an active connection.
+            if connection.is_connected().await {
+                return true;
+            }
+            //  Proceed to remove subscription if found but not connected/no connection..
+            guard.scripthash_subs.remove(script_hash);
+        };
+
+        false
+    }
+
+    /// remove scripthash subscription from list by server addr
+    async fn remove_subscription_by_addr(&self, server_addr: &str) {
+        let mut guard = self.0.state_guard.lock().await;
+        // remove server from scripthash subscription list
+        for (script, conn) in guard.scripthash_subs.clone() {
+            if conn.lock().await.addr == server_addr {
+                guard.scripthash_subs.remove(&script);
+            }
+        }
+    }
+}
+
+impl ConnectionManagerMultiple {
+    async fn suspend_server(&self, address: &str) -> Result<(), ConnectionManagerErr> {
+        debug!(
+            "About to suspend connection to addr: {}, guard: {:?}",
+            address, self.0.state_guard
+        );
+        let mut guard = self.0.state_guard.lock().await;
+
+        Self::reset_connection_context(&mut guard, address, self.0.abortable_system.create_subsystem().unwrap())?;
+
+        let suspend_timeout_sec = Self::get_suspend_timeout(&guard, address).await?;
+        Self::duplicate_suspend_timeout(&mut guard, address)?;
+        drop(guard);
+
+        self.clone().spawn_resume_server(address, suspend_timeout_sec);
+        debug!("Suspend future spawned");
+        Ok(())
+    }
+
+    // workaround to avoid the cycle detected compilation error that blocks recursive async calls
+    fn spawn_resume_server(self, address: &str, suspend_timeout_sec: u64) {
+        let spawner = self.0.abortable_system.weak_spawner();
+        let address = address.to_owned();
+        spawner.spawn(Box::new(
+            async move {
+                debug!("Suspend server: {}, for: {} seconds", address, suspend_timeout_sec);
+                Timer::sleep(suspend_timeout_sec as f64).await;
+                let _ = self.resume_server(&address).await;
+            }
+            .boxed(),
+        ));
+    }
+
+    async fn resume_server(self, address: &str) -> Result<(), ConnectionManagerErr> {
+        debug!("Resume address: {}", address);
+        let guard = self.0.state_guard.lock().await;
+
+        let (_, conn_ctx) = Self::get_connection_ctx(&guard, address)?;
+        let conn_settings = conn_ctx.conn_settings.clone();
+        let conn_spawner = conn_ctx.abortable_system.weak_spawner();
+        drop(guard);
+
+        if let Err(err) = self.clone().connect_to(&conn_settings, conn_spawner).await {
+            error!("Failed to resume: {}", err);
+            self.suspend_server(address).await?;
+        }
```

Comment:
> connect_to probably never fails in the first place because the timeout branch doesn't yield an error:

`connect_to` attempts to establish a connection to an address specified by `conn_settings`. If the connection process exceeds the timeout, then an error is returned indicating a timeout occurred while trying to connect to the server. 

>I am kinda confused of how connect_to calls suspend_server if it fails internally and then we do additional suspend_server when it fails again (which theoretically shouldn't happen).

the `suspend_server` function is not directly called upon connection failure in the `connect_to` as `connect_to` might return other early errors before even reaching the `select!` block.

so I suppose, the `suspend_server` function is called again in the `resume_server` function as a fallback mechanism


cc. @rozhkovdmitrii